File: frontend/src/types/platforms.ts
/**
 * Platform management. Mirrors the backend `PlatformsResponse`.
 *
 * Assembled server-side from the source registry joined to this user's sessions, so Sources
 * and Settings cannot disagree about what the product supports or what is connected — the
 * defect this replaced had Settings hard-coding four platforms against a registry of 55.
 *
 * `actions` carries availability and a reason, because whether a source needs a login is a
 * property of its adapter, not something the browser can work out.
 */

export interface PlatformAction {
  key: 'toggle' | 'connect' | 'disconnect' | 'test' | string;
  label: string;
  available: boolean;
  /** Why it is unavailable. Empty when it is available. */
  reason: string;
}

export interface PlatformStatus {
  key: string;
  label: string;
  tier: string;
  health: string;
  implemented: boolean;
  needs_credential: boolean;
  capabilities: string[];
  enabled: boolean;
  connected: boolean;
  connection_state: string;
  /** Masked, e.g. `sur***@example.com`. Never a credential. */
  account: string | null;
  last_used_at: string | null;
  expires_at: string | null;
  detail: string | null;
  last_error: string | null;
  actions: PlatformAction[];
}

export interface PlatformsResponse {
  platforms: PlatformStatus[];
  total: number;
  /** Sources with a working adapter — the number that matters for discovery. */
  usable: number;
  connected: number;
}

/**
 * An in-flight interactive login capture.
 *
 * The operator signs in on the platform's own page in a real browser window; only the
 * resulting session is kept. Nothing here ever carries a cookie or a credential — this is
 * purely progress, so the UI can say what is happening while a window is open.
 */
export interface ConnectAttempt {
  id: string;
  platform: string;
  state:
    | 'opening'
    | 'awaiting_login'
    | 'capturing'
    | 'connected'
    | 'failed'
    | 'cancelled'
    | 'timed_out';
  /** True once the attempt has settled; the UI stops polling. */
  done: boolean;
  /** What the operator should do right now. */
  instructions: string;
  /** What happened, once it has. */
  detail: string;
  seconds_remaining: number;
}

/** One live probe result — what happened just now, not what the catalogue claims. */
export interface PlatformTestResult {
  key: string;
  label: string;
  state: string;
  /** The upstream's own words on failure. */
  detail: string;
  results: number;
  elapsed_ms: number;
  ok: boolean;
}

export interface PlatformTestResponse {
  results: PlatformTestResult[];
  tested: number;
  passed: number;
  failed: number;
  elapsed_ms: number;
}

File: backend/app/schemas/platforms.py
"""Wire format for platform management.

One shape, assembled from the source registry and the user's stored sessions, so the Sources
screen and the Settings platform section cannot disagree about what the product supports or
what is connected. Two screens maintaining their own lists is how "Sources says Reed is
connected, Settings says it is not" happens.

Nothing here carries a credential. ``account`` is a masked label; the encrypted cookies live
in ``user_credentials`` and are never read on this path.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class PlatformAction(BaseModel):
    """An action the UI may offer for this platform, and whether it is available now.

    ``available`` is decided server-side because the frontend cannot know whether a session
    import is meaningful for a keyless public API. A disabled button with a reason beats a
    button that does nothing.
    """

    key: str
    label: str
    available: bool
    #: Why it is unavailable. Empty when it is.
    reason: str = ""


class PlatformStatus(BaseModel):
    """One source, its capabilities, and this user's connection to it."""

    key: str
    label: str
    tier: str
    #: Registry health: live / degraded / auth_required / not_implemented / blocked …
    health: str
    #: False when the source is catalogued but no adapter can serve it.
    implemented: bool
    #: True when the source cannot run without the operator supplying something.
    needs_credential: bool = False
    #: What this adapter can do — search, application_url, apply, and so on.
    capabilities: list[str] = Field(default_factory=list)

    #: Whether the operator has this source switched on for discovery.
    enabled: bool = False

    # -- Connection, for platforms that need a login -------------------------------------
    #: True only when a session exists AND is in a usable state.
    connected: bool = False
    #: Session lifecycle: session_active, session_expired, mfa_required, not_connected …
    connection_state: str = "not_connected"
    #: Masked, e.g. ``sur***@example.com``. Never a full address, never a credential.
    account: str | None = None
    last_used_at: datetime | None = None
    expires_at: datetime | None = None
    #: Why the connection needs attention, shown verbatim.
    detail: str | None = None

    #: Last time this source returned results, and what went wrong if it did not. Both are
    #: ``None`` when the source has never been exercised — different from "it failed".
    last_error: str | None = None

    actions: list[PlatformAction] = Field(default_factory=list)


class PlatformsResponse(BaseModel):
    """Every source the product knows about, with this user's state attached."""

    platforms: list[PlatformStatus] = Field(default_factory=list)
    total: int = 0
    #: Sources with a working adapter — the number that actually matters for discovery.
    usable: int = 0
    connected: int = 0
